File: src/artifacts/system_health.py
"""Build one multi-watermark health snapshot from governed read models."""

from __future__ import annotations

import json
import os
from datetime import datetime
from pathlib import Path
from typing import Any, Mapping, Sequence

from src.artifacts.model_run_bundle_v2 import validate_catalog
from src.governance.active_strategy_catalog import (
    ActiveStrategy,
    assert_formal_catalog_matches_active_strategies,
    load_active_strategy_catalog,
)

SCHEMA_VERSION = "1.0.0"
STATES = {"current", "delayed", "blocked", "inconsistent", "not_applicable"}
HEALTH_PRECEDENCE = {
    "not_applicable": 0,
    "current": 1,
    "delayed": 2,
    "blocked": 3,
    "inconsistent": 4,
}


class SystemHealthError(ValueError):
    """Raised when governed health inputs are structurally inconsistent."""


def _object(path: Path) -> dict[str, Any]:
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise SystemHealthError(f"invalid health input: {path}") from exc
    if not isinstance(payload, dict):
        raise SystemHealthError(f"health input root must be an object: {path}")
    return payload


def _date(value: object) -> str | None:
    if not isinstance(value, str) or not value:
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00")).date().isoformat()
    except ValueError:
        try:
            return datetime.strptime(value, "%Y-%m-%d").date().isoformat()
        except ValueError as exc:
            raise SystemHealthError(f"invalid health watermark: {value}") from exc


def _max_date(values: Sequence[str | None]) -> str | None:
    retained = [value for value in values if value]
    return max(retained) if retained else None


def _state_max(states: Sequence[str]) -> str:
    invalid = set(states) - STATES
    if invalid:
        raise SystemHealthError(f"unsupported health states: {sorted(invalid)}")
    return max(states, key=lambda value: HEALTH_PRECEDENCE[value])


def _factor_cutoff(operation: Mapping[str, Any]) -> str | None:
    rows = operation.get("factor_evidence")
    if isinstance(rows, list):
        observed = [
            _date(row.get("observed_at"))
            for row in rows
            if isinstance(row, Mapping)
        ]
        if any(observed):
            return _max_date(observed)
    return _date(operation.get("latest_completed_session"))


def _last_signal_change(strategy: ActiveStrategy, repository_root: Path) -> str | None:
    records_root = repository_root / strategy.signal_ledger / "records"
    if not records_root.is_dir():
        return None
    changed: list[str] = []
    for path in sorted(records_root.glob("*.json")):
        payload = _object(path)
        signal = payload.get("signal")
        if not isinstance(signal, Mapping) or signal.get("should_alert") is not True:
            continue
        observed = _date(payload.get("signal_date") or signal.get("signal_date"))
        if observed:
            changed.append(observed)
    return max(changed) if changed else None


def _delivery_state(status_value: object) -> tuple[str, str | None]:
    status = str(status_value or "")
    if status in {"sent"}:
        return "current", status
    if status in {"not_required", "unchanged", "suppressed"}:
        return "not_applicable", status
    if status in {"failed", "delivery_failed", "error"}:
        return "blocked", status
    if status in {"pending", "skipped_not_configured"}:
        return "delayed", status
    if status in {"not available", ""}:
        return "not_applicable", None
    raise SystemHealthError(f"unsupported delivery status: {status}")


def _operation_state(operation: Mapping[str, Any]) -> str:
    status = str(operation.get("status") or "")
    if status in {"blocked", "pipeline_unavailable", "delivery_failed"}:
        return "blocked"
    if status == "stale":
        return "delayed"
    if status in {
        "awaiting_observation",
        "current_no_change",
        "target_pending_execution",
        "execution_observed",
    }:
        return "current"
    raise SystemHealthError(f"unsupported Strategy Operations status: {status}")


def _freshness_state(value: object) -> str:
    text = str(value or "unknown")
    if text == "current":
        return "current"
    if text == "stale":
        return "delayed"
    if text in {"blocked", "unknown"}:
        return "blocked"
    raise SystemHealthError(f"unsupported freshness status: {text}")


def build_system_health(
    *,
    repository_root: Path,
    formal_catalog: Path,
    formal_freshness: Path,
    operations: Mapping[str, Any],
    model_data_readiness: Path,
    generated_at: str,
) -> dict[str, Any]:
    """Build health from the canonical Strategy Operations read model."""

    root = repository_root.resolve()
    active = load_active_strategy_catalog(root / "configs/strategies/registry.json")
    formal = _object(formal_catalog)
    validate_catalog(formal)
    assert_formal_catalog_matches_active_strategies(formal, active)
    freshness = _object(formal_freshness)
    model_data = _object(model_data_readiness)

    operation_rows = operations.get("records")
    formal_rows = formal.get("records")
    if not isinstance(operation_rows, list) or not isinstance(formal_rows, list):
        raise SystemHealthError("formal/operations records are missing")
    operations_by_model = {
        str(row.get("model_version_id")): row
        for row in operation_rows
        if isinstance(row, Mapping)
    }
    formal_by_model = {
        str(row.get("model_version_id")): row
        for row in formal_rows
        if isinstance(row, Mapping)
    }
    expected_ids = set(active.active_model_version_ids)
    if set(operations_by_model) != expected_ids or set(formal_by_model) != expected_ids:
        raise SystemHealthError("health inputs do not exactly match the active strategy set")

    market_cutoffs = freshness.get("markets")
    if not isinstance(market_cutoffs, Mapping):
        raise SystemHealthError("formal provider-resolved market cutoffs are missing")
    model_data_cutoff = _date(model_data.get("evidence_cutoff"))
    model_data_summary = model_data.get("summary")
    if not isinstance(model_data_summary, Mapping):
        raise SystemHealthError("model-data readiness summary is missing")
    model_data_state = (
        "current"
        if int(model_data_summary.get("blocked_component_count", 0)) == 0
        and int(model_data_summary.get("partial_component_count", 0)) == 0
        else "blocked"
    )

    expected_market_candidates: dict[str, list[str | None]] = {}
    last_changes: dict[str, str | None] = {}
    for strategy in active.strategies:
        operation = operations_by_model[strategy.model_version_id]
        formal_record = formal_by_model[strategy.model_version_id]
        last_changes[strategy.model_version_id] = _last_signal_change(strategy, root)
        expected_market_candidates.setdefault(strategy.market, []).extend(
            [
                _date(market_cutoffs.get(strategy.market)),
                _date(formal_record.get("evidence_cutoff")),
                _date(operation.get("latest_completed_session")),
                _date(operation.get("as_of")),
            ]
        )

    markets: list[dict[str, Any]] = []
    market_state: dict[str, str] = {}
    market_expected: dict[str, str | None] = {}
    for market in sorted(expected_market_candidates):
        provider_cutoff = _date(market_cutoffs.get(market))
        expected_cutoff = _max_date(expected_market_candidates[market])
        if provider_cutoff is None or expected_cutoff is None:
            state = "blocked"
        elif provider_cutoff < expected_cutoff:
            state = "delayed"
        elif provider_cutoff == expected_cutoff:
            state = "current"
        else:
            state = "inconsistent"
        market_state[market] = state
        market_expected[market] = expected_cutoff
        markets.append(
            {
                "market": market,
                "state": state,
                "market_expected_cutoff": expected_cutoff,
                "market_expected_cutoff_source": "max_governed_active_watermark",
                "provider_cutoff": provider_cutoff,
                "provider_cutoff_source": "provider_resolved_common_session",
                "provider_lag_sessions": 0 if state == "current" else None,
                "provider_lag_exact": state == "current",
                "provider_formal_consistency": (
                    "current" if state in {"current", "delayed"} else state
                ),
            }
        )

    strategies: list[dict[str, Any]] = []
    for strategy in active.strategies:
        model_id = strategy.model_version_id
        operation = operations_by_model[model_id]
        formal_record = formal_by_model[model_id]
        provider_cutoff = _date(market_cutoffs.get(strategy.market))
        formal_cutoff = _date(formal_record.get("evidence_cutoff"))
        factor_cutoff = _factor_cutoff(operation)
        signal_evaluation = _date(operation.get("as_of")) or _date(
            operation.get("latest_completed_session")
        )
        delivery_state, delivery_status = _delivery_state(
            operation.get("delivery_status")
        )
        formal_state = (
            "blocked"
            if formal_cutoff is None or provider_cutoff is None
            else ("delayed" if formal_cutoff < provider_cutoff else "current")
        )
        data_state = model_data_state
        if (
            data_state == "current"
            and provider_cutoff is not None
            and model_data_cutoff is not None
            and model_data_cutoff < provider_cutoff
        ):
            data_state = "delayed"
        factor_state = _freshness_state(operation.get("factor_freshness"))
        signal_state = _operation_state(operation)
        internal_state = _state_max(
            [market_state[strategy.market], formal_state, data_state, factor_state, signal_state]
        )
        strategies.append(
            {
                "strategy_id": strategy.strategy_id,
                "model_version_id": model_id,
                "market": strategy.market,
                "state": internal_state,
                "market_expected_cutoff": market_expected[strategy.market],
                "provider_cutoff": provider_cutoff,
                "formal_cutoff": formal_cutoff,
                "model_data_cutoff": model_data_cutoff,
                "factor_cutoff": factor_cutoff,
                "last_signal_evaluation": signal_evaluation,
                "last_signal_change": last_changes[model_id],
                "delivery_state": delivery_state,
                "delivery_status": delivery_status,
                "stages": {
                    "provider": market_state[strategy.market],
                    "formal": formal_state,
                    "model_data": data_state,
                    "factor": factor_state,
                    "signal": signal_state,
                    "delivery": delivery_state,
                },
                "formal_bundle_id": formal_record.get("bundle_id"),
                "formal_run_id": formal_record.get("run_id"),
            }
        )

    deployment_commit = os.environ.get("GITHUB_SHA") or None
    deployment_run = os.environ.get("GITHUB_RUN_ID") or None
    return {
        "schema_version": SCHEMA_VERSION,
        "generated_at": generated_at,
        "state": _state_max([row["state"] for row in strategies]),
        "markets": markets,
        "strategies": strategies,
        "deployment": {
            "state": "not_applicable",
            "expected_commit_sha": deployment_commit,
            "workflow_run_id": deployment_run,
            "live_acceptance": "verified_after_deployment",
            "receipt": "deployment.json",
        },
        "model_data": {
            "state": model_data_state,
            "evidence_cutoff": model_data_cutoff,
            "bundle_id": model_data.get("bundle_id"),
        },
        "research_only": True,
        "trade_ready": False,
    }


def validate_system_health(payload: Mapping[str, Any]) -> None:
    if payload.get("schema_version") != SCHEMA_VERSION:
        raise SystemHealthError("unsupported system health schema")
    if payload.get("research_only") is not True or payload.get("trade_ready") is not False:
        raise SystemHealthError("system health research boundary is invalid")
    if payload.get("state") not in STATES:
        raise SystemHealthError("system health state is invalid")
    markets = payload.get("markets")
    strategies = payload.get("strategies")
    if not isinstance(markets, list) or not markets:
        raise SystemHealthError("system health markets are missing")
    if not isinstance(strategies, list) or not strategies:
        raise SystemHealthError("system health strategies are missing")
    for row in [*markets, *strategies]:
        if not isinstance(row, Mapping) or row.get("state") not in STATES:
            raise SystemHealthError("system health row state is invalid")
    for row in strategies:
        stages = row.get("stages")
        if not isinstance(stages, Mapping):
            raise SystemHealthError("strategy health stages are missing")
        if any(value not in STATES for value in stages.values()):
            raise SystemHealthError("strategy stage state is invalid")


def write_system_health(path: Path, payload: Mapping[str, Any]) -> bool:
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")) + "\n"
    current = path.read_text(encoding="utf-8") if path.is_file() else None
    if current == encoded:
        return False
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(encoded, encoding="utf-8")
    return True

File: qlib-dashboard/src/lib/system-health.ts
export type SystemHealthState = 'current' | 'delayed' | 'blocked' | 'inconsistent' | 'not_applicable';

export interface SystemHealthMarket {
  market: string;
  state: SystemHealthState;
  market_expected_cutoff: string | null;
  market_expected_cutoff_source: string;
  provider_cutoff: string | null;
  provider_cutoff_source: string;
  provider_lag_sessions: number | null;
  provider_lag_exact: boolean;
  provider_formal_consistency: SystemHealthState;
}

export interface SystemHealthStrategy {
  strategy_id: string;
  model_version_id: string;
  market: string;
  state: SystemHealthState;
  market_expected_cutoff: string | null;
  provider_cutoff: string | null;
  formal_cutoff: string | null;
  model_data_cutoff: string | null;
  factor_cutoff: string | null;
  last_signal_evaluation: string | null;
  last_signal_change: string | null;
  delivery_state: SystemHealthState;
  delivery_status: string | null;
  stages: Record<'provider' | 'formal' | 'model_data' | 'factor' | 'signal' | 'delivery', SystemHealthState>;
  formal_bundle_id: string;
  formal_run_id: string;
}

export interface SystemHealthSnapshot {
  schema_version: '1.0.0';
  generated_at: string;
  state: SystemHealthState;
  markets: SystemHealthMarket[];
  strategies: SystemHealthStrategy[];
  deployment: {
    state: SystemHealthState;
    expected_commit_sha: string | null;
    workflow_run_id: string | null;
    live_acceptance: string;
    receipt: 'deployment.json';
  };
  model_data: {
    state: SystemHealthState;
    evidence_cutoff: string | null;
    bundle_id: string | null;
  };
  research_only: true;
  trade_ready: false;
}

export interface LiveSystemHealth {
  health: SystemHealthSnapshot | null;
  status: SystemHealthState | 'unknown';
  deploymentStatus: SystemHealthState | 'unknown';
  message: string;
}

const STATES = new Set<SystemHealthState>(['current', 'delayed', 'blocked', 'inconsistent', 'not_applicable']);

function isObject(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
}

function state(value: unknown, label: string): SystemHealthState {
  if (typeof value !== 'string' || !STATES.has(value as SystemHealthState)) {
    throw new Error(`Invalid system-health state: ${label}.`);
  }
  return value as SystemHealthState;
}

function nullableString(value: unknown): string | null {
  return typeof value === 'string' && value.length > 0 ? value : null;
}

export function parseSystemHealth(value: unknown): SystemHealthSnapshot {
  if (!isObject(value)
    || value.schema_version !== '1.0.0'
    || value.research_only !== true
    || value.trade_ready !== false
    || !Array.isArray(value.markets)
    || !Array.isArray(value.strategies)
    || !isObject(value.deployment)
    || !isObject(value.model_data)) {
    throw new Error('System health contract is invalid.');
  }
  const markets = value.markets.map((item, index) => {
    if (!isObject(item)) throw new Error(`System health market ${index} is invalid.`);
    return {
      market: String(item.market ?? ''),
      state: state(item.state, `market ${index}`),
      market_expected_cutoff: nullableString(item.market_expected_cutoff),
      market_expected_cutoff_source: String(item.market_expected_cutoff_source ?? ''),
      provider_cutoff: nullableString(item.provider_cutoff),
      provider_cutoff_source: String(item.provider_cutoff_source ?? ''),
      provider_lag_sessions: typeof item.provider_lag_sessions === 'number' ? item.provider_lag_sessions : null,
      provider_lag_exact: item.provider_lag_exact === true,
      provider_formal_consistency: state(item.provider_formal_consistency, `market consistency ${index}`),
    };
  });
  const strategies = value.strategies.map((item, index) => {
    if (!isObject(item) || !isObject(item.stages)) throw new Error(`System health strategy ${index} is invalid.`);
    const stages = item.stages;
    return {
      strategy_id: String(item.strategy_id ?? ''),
      model_version_id: String(item.model_version_id ?? ''),
      market: String(item.market ?? ''),
      state: state(item.state, `strategy ${index}`),
      market_expected_cutoff: nullableString(item.market_expected_cutoff),
      provider_cutoff: nullableString(item.provider_cutoff),
      formal_cutoff: nullableString(item.formal_cutoff),
      model_data_cutoff: nullableString(item.model_data_cutoff),
      factor_cutoff: nullableString(item.factor_cutoff),
      last_signal_evaluation: nullableString(item.last_signal_evaluation),
      last_signal_change: nullableString(item.last_signal_change),
      delivery_state: state(item.delivery_state, `delivery ${index}`),
      delivery_status: nullableString(item.delivery_status),
      stages: {
        provider: state(stages.provider, `provider ${index}`),
        formal: state(stages.formal, `formal ${index}`),
        model_data: state(stages.model_data, `model_data ${index}`),
        factor: state(stages.factor, `factor ${index}`),
        signal: state(stages.signal, `signal ${index}`),
        delivery: state(stages.delivery, `delivery ${index}`),
      },
      formal_bundle_id: String(item.formal_bundle_id ?? ''),
      formal_run_id: String(item.formal_run_id ?? ''),
    };
  });
  if (markets.length === 0 || strategies.length === 0) throw new Error('System health records are empty.');
  return {
    schema_version: '1.0.0',
    generated_at: String(value.generated_at ?? ''),
    state: state(value.state, 'root'),
    markets,
    strategies,
    deployment: {
      state: state(value.deployment.state, 'deployment'),
      expected_commit_sha: nullableString(value.deployment.expected_commit_sha),
      workflow_run_id: nullableString(value.deployment.workflow_run_id),
      live_acceptance: String(value.deployment.live_acceptance ?? ''),
      receipt: 'deployment.json',
    },
    model_data: {
      state: state(value.model_data.state, 'model_data'),
      evidence_cutoff: nullableString(value.model_data.evidence_cutoff),
      bundle_id: nullableString(value.model_data.bundle_id),
    },
    research_only: true,
    trade_ready: false,
  };
}

function assetUrl(path: string): string {
  const base = import.meta.env.BASE_URL.endsWith('/') ? import.meta.env.BASE_URL : `${import.meta.env.BASE_URL}/`;
  return `${base}${path}`;
}

async function deploymentState(health: SystemHealthSnapshot): Promise<SystemHealthState | 'unknown'> {
  if (!health.deployment.expected_commit_sha) return 'not_applicable';
  try {
    const response = await fetch(assetUrl('deployment.json'), { cache: 'no-cache' });
    if (!response.ok) return 'unknown';
    const receipt = await response.json() as unknown;
    if (!isObject(receipt) || typeof receipt.commit_sha !== 'string') return 'inconsistent';
    return receipt.commit_sha === health.deployment.expected_commit_sha ? 'current' : 'inconsistent';
  } catch {
    return 'unknown';
  }
}

export async function fetchSystemHealth(): Promise<LiveSystemHealth> {
  try {
    const response = await fetch(assetUrl('data/strategy-operations/system-health.json'), { cache: 'no-cache' });
    if (!response.ok) {
      return { health: null, status: 'unknown', deploymentStatus: 'unknown', message: `System health is unavailable (${response.status}).` };
    }
    const health = parseSystemHealth(await response.json() as unknown);
    const deploymentStatus = await deploymentState(health);
    const status = health.state;
    const delayed = health.markets.filter((row) => row.state === 'delayed').map((row) => row.market.toUpperCase());
    const message = status === 'current'
      ? 'Provider, model-data, factor, formal, signal and delivery stages are current.'
      : status === 'delayed'
        ? `Pipeline is internally consistent but delayed${delayed.length ? ` in ${delayed.join(' / ')}` : ''}.`
        : status === 'blocked'
          ? 'At least one governed pipeline stage is blocked.'
          : status === 'inconsistent'
            ? 'Governed pipeline watermarks disagree; publication should fail closed.'
            : 'No runtime health evaluation applies to this build.';
    return { health, status, deploymentStatus, message };
  } catch (error) {
    return { health: null, status: 'unknown', deploymentStatus: 'unknown', message: error instanceof Error ? error.message : 'System health request failed.' };
  }
}
